File: Utilities/DatabaseUpdater/UpdateScript.vb
Public Class UpdateScript
    Inherits Script
    Implements IComparable
	Friend Const ExtendedPropertyForVersionInfoName As String = "DatabaseVersion|ModifiedDate|UpdateTimestamp|MostRecentScriptName"
	Private _versionNumber As Integer
	Sub New(ByVal path As String)
		MyBase.New(path)
		Dim versionNumberString As String = Me.GetFileName.Substring(0, Me.GetFileName.IndexOf("-")).Trim
		Me._versionNumber = Integer.Parse(versionNumberString)
		Me.objectType = ScriptType.UpdateScript
	End Sub


	Public Overrides Function ShouldBeApplied() As Boolean
		Try
			Try
				Dim valueOfExtendedPropertyForVersionInfoName As String = SQLExtendedPropertyMethods.Get(UpdateScript.ExtendedPropertyForVersionInfoName)
				Dim databaseVersionNumber As Integer = Convert.ToInt32(valueOfExtendedPropertyForVersionInfoName.Split("|"c)(0))
				Dim modifiedDateOfMostRecentScript As DateTime = DateTime.Parse(valueOfExtendedPropertyForVersionInfoName.Split("|"c)(1))
				If Me.VersionNumber < databaseVersionNumber Then
					Return False
				ElseIf Me.VersionNumber = databaseVersionNumber Then
					Return Me.modified.CompareTo(modifiedDateOfMostRecentScript) > 0
				Else
					Return True
				End If
			Catch ex As NoDatabaseObjectWithThatNameException
				Return True
			End Try
		Catch ex As Exception
			Throw ex
		End Try
	End Function
	Friend ReadOnly Property VersionNumber() As Integer
		Get
			Return Me._versionNumber
		End Get
	End Property


	Public Function CompareTo(ByVal obj As Object) As Integer Implements System.IComparable.CompareTo
		Return Me._versionNumber.CompareTo(CType(obj, UpdateScript).VersionNumber)
	End Function
	Public Overrides Sub LogUpdateInDatabase()
		SQLExtendedPropertyMethods.Update(UpdateScript.ExtendedPropertyForVersionInfoName, Me.VersionNumber & "|" & Me.modified.ToString() & "|" & DateTime.Now & "|" & Me.GetFileName)
	End Sub
	
End Class

File: Utilities/SqlScriptRunner/ScriptTypes/UpdateScript.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Common.Automation.Sql;
namespace SqlScriptRunner.ScriptTypes
{
    internal  class UpdateScript : Script, IComparable
    {
        internal const string ExtendedPropertyForVersionInfoName = "DatabaseVersion|ModifiedDate|UpdateTimestamp|MostRecentScriptName";
		internal int VersionNumber
		{
			get
			{
				return Int32.Parse(File.Name.Substring(0, File.Name.IndexOf('-')).Trim());
			}
		}
        internal UpdateScript(string path)
            : base(path, SqlScriptRunner.ScriptType.UpdateScript)
        {

            string versionNumberString = this.File.Name.Substring(0, this.File.Name.IndexOf("-")).Trim();
            this.ScriptType = SqlScriptRunner.ScriptType.UpdateScript;
        }

        internal  override bool ShouldBeApplied()
        {
            try
            {
                string valueOfExtendedPropertyForVersionInfoName = SqlExtendedPropertyMethods.GetDatabaseLevelExtendedProperty(Global.GetConnection(), UpdateScript.ExtendedPropertyForVersionInfoName);
                int databaseVersionNumber = Convert.ToInt32(valueOfExtendedPropertyForVersionInfoName.Split('|')[0]);
                
                if (this.VersionNumber < databaseVersionNumber)
                {
                    return false;
                }
                else if (this.VersionNumber == databaseVersionNumber)
                {
                    DateTime modifiedDateOfMostRecentScript = DateTime.Parse(valueOfExtendedPropertyForVersionInfoName.Split('|')[1]);
                    return Math.Abs(this.File.LastWriteTime.Subtract(modifiedDateOfMostRecentScript).TotalSeconds) > 1;
                }
                else
                {
                    return true;
                }
            }
            catch (SqlExtendedPropertyMethods.ObjectDoesNotExist ex)
            {
                return true;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }



        internal  override void LogUpdateInDatabase()
        {
			SqlExtendedPropertyMethods.UpdateDatabaseLevelProperty(Global.GetConnection(), UpdateScript.ExtendedPropertyForVersionInfoName, this.VersionNumber + "|" + this.File.LastWriteTime.ToString() + "|" + DateTime.Now + "|" + this.File.Name);
        }

        #region IComparable Members

        public int CompareTo(object obj)
        {
            return this.VersionNumber.CompareTo(((UpdateScript)(obj)).VersionNumber);
        }

        #endregion
    }
}
